src/compiler/checker.ts

```diff
@@ -3273,6 +3276,10 @@ namespace ts {
                     });
                     extendExportSymbols(symbols, nestedSymbols);
                 }
+                if (symbol.escapedName === InternalSymbolName.ExportEquals && isInJSFile(symbol.valueDeclaration) && isObjectLiteralExpression((symbol.valueDeclaration as BinaryExpression).right)) {
+                    const objectLiteralSymbol = (symbol.valueDeclaration as BinaryExpression).right.symbol;
+                    extendExportSymbols(symbols, objectLiteralSymbol.members);
+                }

```

Comment:
I think this is probably the wrong place for this, but wanted to get some opinions. My first instinct was to make `module.exports = { foo }` recognized by the binder as something special, but that got complicated and blew up very quickly. I could just move this to `importFixes`, but was never really satisfied with having to look so closely at syntax at this point.

Reply:
I'm like 99% sure that if we did something like this, it'd belong in `getCommonJsExportEquals` with our other adhoc cjsy export-assignment code, rather than in `getExportsOfModuleWorker` directly (so the result is cached and considered _before_ any potential reexports).

_However_. Today, `module.exports = { foo }` is no different than `export = { foo }` (and I don't think we should privilege one more than the other behaviorally), and `export = { 
foo }` creates an `export=` member on the containing file which is essentially just a variable declaration. While these members can be looked up as "exports" of the file, they are essentially only "members" of the declaration. The "exports" vs "members" distinction is a very internal thing that, via aliasing, isn't always clear (and is _very unfortunately_ exposed by our public `getExportsOfModule` function). However, what _is_ clear is the uniform way to get the full, final list of exported members of a module, and it's _not_ the checker-internal `getExportsOfModule` function (hilariously enough); that exists to traverse over a (admittedly flattened) representation of the exports-as-written, _without_ then layering on an `export = <expr>` resolution.

Long story short, there's a `checker.getExportsOfModule` call in `importFixes.ts` that should probably be `checker.getExportsAndPropertiesOfModule` instead (the later gets the "finalized" exports by deferring to `getPropertiesOfType(getTypeOfSymbol(moduleSymbol))`). Once you swap that (which affects no existing tests, I just checked), I'm pretty sure you shouldn't need this bit of checker code at all.

Reply:
Hmm, good to know that exists, and I agree there’s no reason to treat `module.exports = { foo }` differently from `exports = { foo }`. But, I’m not sure I want to get members if it’s something like

```ts
exports = new SingletonThing();
```

or something like

```ts
exports = styled.button<ThemeOne | ThemeTwo | ThemeThree | ...>`
  fontFamily: ${props => props.theme.fontFamily}
`;
```

because a) members of those would be nonsensical to auto-import, and b) this code path also happens for completions-list import suggestions, and since `getPropertiesOfType` will happily resolve all members of union and intersection types, I’m concerned that this is ripe for a 10x slowdown for everyone who happens to have `@types/coolest-most-cleverest-types-ever` installed.

I think it probably makes sense only to dive into properties if the `export=` symbol actually resolves to an _object literal_. Where do you think would make the most sense to do that?

Reply:
I don't see why
```ts
const a = {
  foo
};
module.exports = a;
```
is any different, from the perspective of our analysis, from
```ts
module.exports = {
  foo
};
```
- the first is just an alias pointing at the second. Moreover,
```ts
module.exports = (function() {
  return { foo };
})();
```
also isn't a wholly uncommon pattern and is fairly reasonable. I also remember `lighthouse` has a bunch of export assigned call expressions that do things like annotate a class with more members, which is another valid pattern.

Sure, I'd strip any autoimports from _primitives_ (or arrays) that get export-assigned (since who would want an autoimport for a string export's `valueOf`), but that can be filtered for after the fact, IMO. Pretty much any export assigned objecty type produces useful export member names; I don't think the filtering here should be syntactic in nature.

Reply:
Actually, I think a good heuristic is probably just dropping all symbols returned by `getExportsAndPropertiesOfModule` which have a `declaration` in the `lib`... technically this's mean, eg, a `Promise` polyfill which just reexports the global type doesn't get import completions, but that's probably fine the global is available).

Reply:
That’s fair. We already drop members of primitives, arrays, and objects that are definitely class instances (I added the latter two conditions a few months ago, in fact.) And to be fair, I think we would/could easily resolve `module.exports = foo` to `foo`’s declaration to see if _it_ was an object literal, since that’s bound as an alias. But anyway, are you not worried about the potential performance impact of running `getPropertiesOfType` on every `export=` symbol we can find?

Reply:
Pfffffffffft, `getPropertiesOfType` isn't recursive or anything, it's just getting members (and not the types of those members). _Most of the time_, it _is_ just a `symbol.exports` or `symbol.members` lookup; it just also handles all the times when that's _not_ the case (which is when it's an expression), and imo generally they're cases worth handling. It's not worse than calling `checkExpressionCached` on every `export = <expr>` position (since that's all it desugars to in the 'worst' case, essentially).

Reply:
FWIW I'd also be scared to call `checkExpressionCached` on every `export=`, but I’ll give it a shot 😄